tests/test_outputs/test_form_output_prep.py

```diff
@@ -0,0 +1,329 @@
+            [39900008914, 1, 'US ', 0.0, 'E', 17219, 1, 'Yes', 600, 100, "0003", 9.0],
+            [39900008968, 1, 'GB ', 1.44, 'AA', 46420, 1, 'Yes', 600, 100, "0003", 40.0],
+            [39900009016, 1, 'GB ', 4.0, 'AD', 71129, 1, 'Yes', 600, 100, "0003", 50.0],
+            [39900009078, 1, 'GB ', 0.72, 'J', 23610, 1, 'Yes', 600, 100, "0003", 15.0]
+        ]
+        ni_expected = pd.DataFrame(data=data, columns=columns)
+        ni_expected["201"] = ni_expected["201"].astype("category")
+        ni_expected["pg_numeric"] = ni_expected["pg_numeric"].astype("category")
+        return ni_expected
+    
+
+    @pytest.fixture(scope="function")
+    def full_output_expected(self) -> pd.DataFrame:
+        columns = [
+            'reference',
+            'instance',
+            'period',
+            '201',
+            '202',
+            '211',
+            '602',
+            '604',
+            'a_weight',
+            'formtype', 
+            'pg_numeric', 
+            'period_year'
+        ]
+        data = [
+            [39900000404, 0, 202212, 'AA', np.nan, np.nan, 100.0, 'Yes', 6.0, 6, 40, 2022],
+            [39900000404, 1, 202212, 'AA', 3146383.0, 4628363.0, 100.0, 'Yes', 6.0, 6, 40, 2022],
+            [39900000404, 2, 202212, 'AA', 0.0, 0.0, 100.0, 'Yes', 6.0, 6, 40, 2022],
+            [39900000408, 0, 202212, 'AA', np.nan, np.nan, 100.0, 'No', 1.0, 6, 40, 2022],
+            [39900000408, 2, 202212, 'AA', 0.0, 0.0, 100.0, 'No', 1.0, 6, 40, 2022],
+            [39900000576, 1, 202212, 'AA', 0.0, 0.0, 100.0, np.nan, 1.0, 1, 40, 2022],
+            [39900000960, 0, 202212, 'AA', np.nan, np.nan, np.nan, 'No', 1.0, 1, 40, 2022],
+            [39900000960, 1, 202212, 'AA', np.nan, 0.0, np.nan, 'No', 1.0, 1, 40, 2022],
+            [39900001029, 1, 202212, 'I', 119376.5, 244543.0, 100.0, np.nan, 1.0, 1, 14, 2022]
+        ]
+        full_outputs = pd.DataFrame(data=data, columns=columns)
+        return full_outputs
+    
+
+    @pytest.fixture(scope="function")
+    def tau_expected(self) -> pd.DataFrame:
+        columns = [
+            'reference',
+            'instance',
+            'period',
+            '201',
+            '202',
+            '211',
+            '602',
+            '604',
+            'a_weight',
+            'formtype',
+            'pg_numeric',
+            'period_year',
+            'ultfoc',
+            '407',
+            'rusic',
+            'form_status']
+        data = [
+            [39900001031, 2, 202212.0, 'U', 0.0, 0.0, 100.0, 'No', 49.0, 6, 30.0, 2022.0, np.nan, np.nan, np.nan, np.nan],
+            [39900001032, 1, 202212.0, 'I', 119376.5, 244543.0, 100.0, np.nan, 1.0, 1, 14.0, 2022.0, np.nan, np.nan, np.nan, np.nan],
+            [39900001577, 1, np.nan, 'C', np.nan, np.nan, 100.0, 'Yes', 1.0, "0003", 3.0, np.nan, 'GB ', 0.72, 10130.0, 600.0],
+            [39900006060, 1, np.nan, 'G', np.nan, np.nan, 100.0, 'Yes', 1.0, "0003", 12.0, np.nan, 'GB ', 0.6, 20120.0, 600.0],
```

Comment:
please could you update so the period_year is of the form yyyy integer, and the NI records also have year values (have a look at the actual outputs if in doubt). This will mean including the relevant column in your NI input. I understand that the other nulls are because you don't include all columns in your inputs, maybe you could make a note to this effect. Thanks.